File: native/mlx/Sources/ClapMLXCache/CacheIdentity.swift
import ClapCacheBridge
import Foundation

public struct OpaquePhysicalCacheIdentityInput: Decodable, Equatable, Sendable {
  public let fingerprint: String
  public let backend: String
  public let resolvedRevision: String
  public let modelArtifactFingerprint: String
  public let tokenizerFingerprint: String
  public let contextAllocation: Int
  public let kvFormat: String
  public let unifiedKV: Bool
  public let layoutVersion: Int

  enum CodingKeys: String, CodingKey, CaseIterable {
    case fingerprint, backend
    case resolvedRevision = "resolved_revision"
    case modelArtifactFingerprint = "model_artifact_fingerprint"
    case tokenizerFingerprint = "tokenizer_fingerprint"
    case contextAllocation = "context_allocation"
    case kvFormat = "kv_format"
    case unifiedKV = "unified_kv"
    case layoutVersion = "layout_version"
  }

  public init(from decoder: any Decoder) throws {
    let values = try decoder.container(keyedBy: CodingKeys.self)
    try rejectUnknownKeys(decoder, allowed: Set(CodingKeys.allCases.map(\.stringValue)),
      name: "cache_identity.physical")
    fingerprint = try values.decode(String.self, forKey: .fingerprint)
    backend = try values.decode(String.self, forKey: .backend)
    resolvedRevision = try values.decode(String.self, forKey: .resolvedRevision)
    modelArtifactFingerprint = try values.decode(String.self, forKey: .modelArtifactFingerprint)
    tokenizerFingerprint = try values.decode(String.self, forKey: .tokenizerFingerprint)
    contextAllocation = try values.decode(Int.self, forKey: .contextAllocation)
    kvFormat = try values.decode(String.self, forKey: .kvFormat)
    unifiedKV = try values.decode(Bool.self, forKey: .unifiedKV)
    layoutVersion = try values.decode(Int.self, forKey: .layoutVersion)
  }
}

public struct OpaqueCacheIdentityDisplay: Decodable, Equatable, Sendable {
  public let namespace: String?
  public let project: String?
  public let harness: String?
  public let agent: String?
  public let session: String?

  enum CodingKeys: String, CodingKey, CaseIterable {
    case namespace, project, harness, agent, session
  }

  public init(from decoder: any Decoder) throws {
    let values = try decoder.container(keyedBy: CodingKeys.self)
    try rejectUnknownKeys(decoder, allowed: Set(CodingKeys.allCases.map(\.stringValue)),
      name: "cache_identity.display")
    namespace = try values.decodeIfPresent(String.self, forKey: .namespace)
    project = try values.decodeIfPresent(String.self, forKey: .project)
    harness = try values.decodeIfPresent(String.self, forKey: .harness)
    agent = try values.decodeIfPresent(String.self, forKey: .agent)
    session = try values.decodeIfPresent(String.self, forKey: .session)
  }
}

public struct OpaqueCacheIdentityInput: Decodable, Equatable, Sendable {
  public let version: Int
  public let generation: String
  public let tenantRoot: String
  public let projectFingerprint: String?
  public let harnessFingerprint: String?
  public let agentFingerprint: String?
  public let sessionFingerprint: String?
  public let scope: String
  public let scopeFingerprint: String
  public let namespaceFingerprint: String
  public let namespaceID: String
  public let priority: String
  public let sideRequest: Bool
  public let display: OpaqueCacheIdentityDisplay
  public let physical: OpaquePhysicalCacheIdentityInput

  enum CodingKeys: String, CodingKey, CaseIterable {
    case version, generation
    case tenantRoot = "tenant_root"
    case projectFingerprint = "project_fingerprint"
    case harnessFingerprint = "harness_fingerprint"
    case agentFingerprint = "agent_fingerprint"
    case sessionFingerprint = "session_fingerprint"
    case scope
    case scopeFingerprint = "scope_fingerprint"
    case namespaceFingerprint = "namespace_fingerprint"
    case namespaceID = "namespace_id"
    case priority
    case sideRequest = "side_request"
    case display, physical
  }

  public init(from decoder: any Decoder) throws {
    let values = try decoder.container(keyedBy: CodingKeys.self)
    try rejectUnknownKeys(decoder, allowed: Set(CodingKeys.allCases.map(\.stringValue)),
      name: "cache_identity")
    version = try values.decode(Int.self, forKey: .version)
    generation = try values.decode(String.self, forKey: .generation)
    tenantRoot = try values.decode(String.self, forKey: .tenantRoot)
    projectFingerprint = try values.decodeIfPresent(String.self, forKey: .projectFingerprint)
    harnessFingerprint = try values.decodeIfPresent(String.self, forKey: .harnessFingerprint)
    agentFingerprint = try values.decodeIfPresent(String.self, forKey: .agentFingerprint)
    sessionFingerprint = try values.decodeIfPresent(String.self, forKey: .sessionFingerprint)
    scope = try values.decode(String.self, forKey: .scope)
    scopeFingerprint = try values.decode(String.self, forKey: .scopeFingerprint)
    namespaceFingerprint = try values.decode(String.self, forKey: .namespaceFingerprint)
    namespaceID = try values.decode(String.self, forKey: .namespaceID)
    priority = try values.decodeIfPresent(String.self, forKey: .priority) ?? "normal"
    sideRequest = try values.decode(Bool.self, forKey: .sideRequest)
    display = try values.decode(OpaqueCacheIdentityDisplay.self, forKey: .display)
    physical = try values.decode(OpaquePhysicalCacheIdentityInput.self, forKey: .physical)
  }
}

public struct PhysicalCacheDescriptor: Equatable, Sendable {
  public let backend: String
  public let contextAllocation: Int
  public let kvFormat: String
  public let unifiedKV: Bool
  public let layoutVersion: Int

  public init(backend: String, contextAllocation: Int, kvFormat: String,
              unifiedKV: Bool, layoutVersion: Int) {
    self.backend = backend
    self.contextAllocation = contextAllocation
    self.kvFormat = kvFormat
    self.unifiedKV = unifiedKV
    self.layoutVersion = layoutVersion
  }
}

public enum CacheIdentityError: Error, CustomStringConvertible, Equatable {
  case invalid(String)
  public var description: String {
    switch self { case .invalid(let message): return message }
  }
}

public struct CacheIdentity {
  public let fingerprint: [UInt8]
  public let tenant: UInt64
  public let project: UInt64
  public let harness: UInt64
  public let agent: UInt64
  public let session: UInt64
  public let scope: UInt32
  public let priority: UInt32
  public let sideRequest: Bool
  public let exportedNamespace: String?
  public let generation: String
  public let physicalFingerprint: String

  public init(input: OpaqueCacheIdentityInput, expected: PhysicalCacheDescriptor) throws {
    guard input.version == 1 else { throw invalid("cache_identity.version must be 1") }
    try validateString(input.generation, key: "generation", maximum: 64)
    let tenantBytes = try digest(input.tenantRoot, key: "tenant_root")
    let namespaceBytes = try digest(input.namespaceFingerprint, key: "namespace_fingerprint")
    let scopeBytes = try digest(input.scopeFingerprint, key: "scope_fingerprint")
    let projectBytes = try input.projectFingerprint.map { try digest($0, key: "project_fingerprint") }
    let harnessBytes = try input.harnessFingerprint.map { try digest($0, key: "harness_fingerprint") }
    let agentBytes = try input.agentFingerprint.map { try digest($0, key: "agent_fingerprint") }
    let sessionBytes = try input.sessionFingerprint.map { try digest($0, key: "session_fingerprint") }
    guard let namespaceID = UInt64(input.namespaceID), input.namespaceID.first != "0",
          String(namespaceID) == input.namespaceID, namespaceID != 0 else {
      throw invalid("cache_identity.namespace_id must be a nonzero decimal u64")
    }
    guard namespaceID == reduce(namespaceBytes) else {
      throw invalid("cache_identity.namespace_id does not match namespace_fingerprint")
    }
    let selectedScope: ([UInt8]?, UInt32) = switch input.scope {
    case "tenant": (tenantBytes, UInt32(CC_SCOPE_TENANT))
    case "project": (projectBytes, UInt32(CC_SCOPE_PROJECT))
    case "harness": (harnessBytes, UInt32(CC_SCOPE_HARNESS))
    case "agent": (agentBytes, UInt32(CC_SCOPE_AGENT))
    case "session": (sessionBytes, UInt32(CC_SCOPE_SESSION))
    default: throw invalid("cache_identity.scope is invalid")
    }
    guard selectedScope.0 == scopeBytes else {
      throw invalid("\(input.scope) scope fingerprint mismatch")
    }
    let selectedPriority: UInt32 = switch input.priority {
    case "interactive": UInt32(CC_PRIORITY_INTERACTIVE)
    case "normal": UInt32(CC_PRIORITY_NORMAL)
    case "background": UInt32(CC_PRIORITY_BACKGROUND)
    default: throw invalid("cache_identity.priority is invalid")
    }
    for (key, label) in [("namespace", input.display.namespace),
                         ("project", input.display.project), ("harness", input.display.harness),
                         ("agent", input.display.agent), ("session", input.display.session)] {
      if let label { try validateString(label, key: "display.\(key)", maximum: 128) }
    }
    _ = try digest(input.physical.fingerprint, key: "physical.fingerprint")
    _ = try digest(input.physical.modelArtifactFingerprint,
      key: "physical.model_artifact_fingerprint")
    _ = try digest(input.physical.tokenizerFingerprint,
      key: "physical.tokenizer_fingerprint")
    try validateString(input.physical.resolvedRevision,
      key: "physical.resolved_revision", maximum: 256)
    try validateString(input.physical.kvFormat, key: "physical.kv_format", maximum: 64)
    guard input.physical.backend == expected.backend,
          input.physical.contextAllocation == expected.contextAllocation,
          input.physical.kvFormat == expected.kvFormat,
          input.physical.unifiedKV == expected.unifiedKV,
          input.physical.layoutVersion == expected.layoutVersion else {
      throw invalid("cache_identity physical model domain does not match loaded runtime")
    }
    fingerprint = namespaceBytes
    tenant = reduce(tenantBytes)
    project = projectBytes.map(reduce) ?? 0
    harness = harnessBytes.map(reduce) ?? 0
    agent = agentBytes.map(reduce) ?? 0
    session = sessionBytes.map(reduce) ?? 0
    scope = selectedScope.1
    priority = selectedPriority
    sideRequest = input.sideRequest
    exportedNamespace = input.display.namespace
    generation = input.generation
    physicalFingerprint = input.physical.fingerprint
  }
}

private struct AnyCodingKey: CodingKey {
  let stringValue: String
  let intValue: Int? = nil
  init?(stringValue: String) { self.stringValue = stringValue }
  init?(intValue: Int) { return nil }
}

private func rejectUnknownKeys(_ decoder: any Decoder, allowed: Set<String>,
                               name: String) throws {
  let keys = try decoder.container(keyedBy: AnyCodingKey.self).allKeys.map(\.stringValue)
  if let unknown = keys.first(where: { !allowed.contains($0) }) {
    throw DecodingError.dataCorrupted(.init(codingPath: decoder.codingPath,
      debugDescription: "\(name) contains unknown field: \(unknown)"))
  }
}

private func validateString(_ value: String, key: String, maximum: Int) throws {
  guard !value.isEmpty, value.utf8.count <= maximum else {
    throw invalid("cache_identity.\(key) has invalid length")
  }
}

private func digest(_ value: String, key: String) throws -> [UInt8] {
  guard value.utf8.count == 64 else {
    throw invalid("cache_identity.\(key) must be 64 lowercase hex characters")
  }
  let characters = Array(value.utf8)
  var bytes: [UInt8] = []
  bytes.reserveCapacity(32)
  for index in stride(from: 0, to: 64, by: 2) {
    guard let high = nibble(characters[index]), let low = nibble(characters[index + 1]) else {
      throw invalid("cache identity fingerprints must use lowercase hexadecimal")
    }
    bytes.append((high << 4) | low)
  }
  return bytes
}

private func nibble(_ character: UInt8) -> UInt8? {
  switch character {
  case 48...57: character - 48
  case 97...102: character - 97 + 10
  default: nil
  }
}

private func reduce(_ bytes: [UInt8]) -> UInt64 {
  let value = bytes.prefix(8).reduce(UInt64(0)) { ($0 << 8) | UInt64($1) }
  return value == 0 ? 1 : value
}

private func invalid(_ message: String) -> CacheIdentityError {
  .invalid(message)
}

File: native/llama/src/cache-identity.cpp
#include "clap/llama/cache-identity.h"

#include "clap/llama/model-runtime.h"

#include <array>
#include <limits>
#include <optional>
#include <set>

namespace clap::llama {
namespace {

using Digest = std::array<uint8_t, 32>;

void exact_keys(const nlohmann::json& value, const std::set<std::string>& allowed,
                const char* name) {
  if (!value.is_object()) throw CacheIdentityError(std::string(name) + " must be an object");
  for (const auto& [key, unused] : value.items()) {
    (void)unused;
    if (!allowed.count(key)) throw CacheIdentityError(std::string(name) + " contains unknown field: " + key);
  }
}

std::string required_string(const nlohmann::json& value, const char* key,
                            std::size_t maximum = std::numeric_limits<std::size_t>::max()) {
  if (!value.contains(key) || !value[key].is_string()) {
    throw CacheIdentityError(std::string("cache_identity.") + key + " must be a string");
  }
  const std::string result = value[key].get<std::string>();
  if (result.empty() || result.size() > maximum) {
    throw CacheIdentityError(std::string("cache_identity.") + key + " has invalid length");
  }
  return result;
}

Digest digest(const nlohmann::json& value, const char* key) {
  const std::string hex = required_string(value, key, 64);
  if (hex.size() != 64) throw CacheIdentityError(std::string("cache_identity.") + key + " must be 64 lowercase hex characters");
  Digest result{};
  for (std::size_t index = 0; index < result.size(); ++index) {
    const char high = hex[index * 2];
    const char low = hex[index * 2 + 1];
    auto nibble = [](char character) -> uint8_t {
      if (character >= '0' && character <= '9') return static_cast<uint8_t>(character - '0');
      if (character >= 'a' && character <= 'f') return static_cast<uint8_t>(character - 'a' + 10);
      throw CacheIdentityError("cache identity fingerprints must use lowercase hexadecimal");
    };
    result[index] = static_cast<uint8_t>((nibble(high) << 4) | nibble(low));
  }
  return result;
}

uint64_t reduce(const Digest& value) {
  uint64_t result = 0;
  for (std::size_t index = 0; index < 8; ++index) result = (result << 8) | value[index];
  return result == 0 ? 1 : result;
}

std::optional<Digest> optional_full_digest(const nlohmann::json& value, const char* key) {
  return value.contains(key) ? std::optional<Digest>(digest(value, key)) : std::nullopt;
}

uint64_t decimal_u64(const std::string& value) {
  if (value.empty() || value[0] == '0') throw CacheIdentityError("cache_identity.namespace_id must be a nonzero decimal u64");
  uint64_t result = 0;
  for (const char character : value) {
    if (character < '0' || character > '9') throw CacheIdentityError("cache_identity.namespace_id must be a nonzero decimal u64");
    const uint64_t digit = static_cast<uint64_t>(character - '0');
    if (result > (UINT64_MAX - digit) / 10) throw CacheIdentityError("cache_identity.namespace_id exceeds u64");
    result = result * 10 + digit;
  }
  if (result == 0) throw CacheIdentityError("cache_identity.namespace_id must be nonzero");
  return result;
}

std::string display(const nlohmann::json& value, const char* key) {
  if (!value.contains(key)) return "";
  if (!value[key].is_string() || value[key].get_ref<const std::string&>().empty() ||
      value[key].get_ref<const std::string&>().size() > 128) {
    throw CacheIdentityError(std::string("cache_identity.display.") + key + " is invalid");
  }
  return value[key].get<std::string>();
}

}  // namespace

PhysicalCacheDescriptor physical_cache_descriptor(const ModelRuntime& runtime) {
  return {"llama", runtime.backend_allocation_cap(), runtime.kv_format(),
          runtime.unified_kv(), 1};
}

ParsedCacheIdentity parse_cache_identity(const nlohmann::json& value,
                                         const ModelRuntime& runtime) {
  return parse_cache_identity(value, physical_cache_descriptor(runtime));
}

ParsedCacheIdentity parse_cache_identity(const nlohmann::json& value,
                                         const PhysicalCacheDescriptor& expected) {
  exact_keys(value, {"version", "generation", "tenant_root", "project_fingerprint",
      "harness_fingerprint", "agent_fingerprint", "session_fingerprint", "scope",
      "scope_fingerprint", "namespace_fingerprint", "namespace_id", "priority",
      "side_request", "display", "physical"}, "cache_identity");
  if (!value.contains("version") || !value["version"].is_number_integer() ||
      value["version"].get<int64_t>() != 1) throw CacheIdentityError("cache_identity.version must be 1");
  ParsedCacheIdentity result;
  result.generation = required_string(value, "generation", 64);
  const Digest tenant = digest(value, "tenant_root");
  const Digest scope_digest = digest(value, "scope_fingerprint");
  const Digest namespace_digest = digest(value, "namespace_fingerprint");
  const auto project = optional_full_digest(value, "project_fingerprint");
  const auto harness = optional_full_digest(value, "harness_fingerprint");
  const auto agent = optional_full_digest(value, "agent_fingerprint");
  const auto session = optional_full_digest(value, "session_fingerprint");
  result.authority.name_space = namespace_digest;
  result.authority.tenant = reduce(tenant);
  result.authority.project = project ? reduce(*project) : 0;
  result.authority.harness = harness ? reduce(*harness) : 0;
  result.authority.agent = agent ? reduce(*agent) : 0;
  result.authority.session = session ? reduce(*session) : 0;
  if (decimal_u64(required_string(value, "namespace_id", 20)) != reduce(namespace_digest)) {
    throw CacheIdentityError("cache_identity.namespace_id does not match namespace_fingerprint");
  }
  const std::string scope = required_string(value, "scope", 16);
  if (scope == "tenant") { result.authority.scope = CLAP_CACHE_SCOPE_TENANT; if (scope_digest != tenant) throw CacheIdentityError("tenant scope fingerprint mismatch"); }
  else if (scope == "project") { result.authority.scope = CLAP_CACHE_SCOPE_PROJECT; if (!project || scope_digest != *project) throw CacheIdentityError("project scope fingerprint mismatch"); }
  else if (scope == "harness") { result.authority.scope = CLAP_CACHE_SCOPE_HARNESS; if (!harness || scope_digest != *harness) throw CacheIdentityError("harness scope fingerprint mismatch"); }
  else if (scope == "agent") { result.authority.scope = CLAP_CACHE_SCOPE_AGENT; if (!agent || scope_digest != *agent) throw CacheIdentityError("agent scope fingerprint mismatch"); }
  else if (scope == "session") { result.authority.scope = CLAP_CACHE_SCOPE_SESSION; if (!session || scope_digest != *session) throw CacheIdentityError("session scope fingerprint mismatch"); }
  else throw CacheIdentityError("cache_identity.scope is invalid");
  const std::string priority = value.contains("priority")
      ? required_string(value, "priority", 16) : "normal";
  if (priority == "interactive") result.authority.priority = CLAP_CACHE_PRIORITY_INTERACTIVE;
  else if (priority == "normal") result.authority.priority = CLAP_CACHE_PRIORITY_NORMAL;
  else if (priority == "background") result.authority.priority = CLAP_CACHE_PRIORITY_BACKGROUND;
  else throw CacheIdentityError("cache_identity.priority is invalid");
  if (!value.contains("side_request") || !value["side_request"].is_boolean()) throw CacheIdentityError("cache_identity.side_request must be boolean");
  result.authority.side_request = value["side_request"].get<bool>();

  if (!value.contains("display")) throw CacheIdentityError("cache_identity.display is required");
  exact_keys(value["display"], {"namespace", "project", "harness", "agent", "session"}, "cache_identity.display");
  result.display = {display(value["display"], "namespace"), display(value["display"], "project"),
      display(value["display"], "harness"), display(value["display"], "agent"),
      display(value["display"], "session")};

  if (!value.contains("physical")) throw CacheIdentityError("cache_identity.physical is required");
  const auto& physical = value["physical"];
  exact_keys(physical, {"fingerprint", "backend", "resolved_revision",
      "model_artifact_fingerprint", "tokenizer_fingerprint", "context_allocation",
      "kv_format", "unified_kv", "layout_version"}, "cache_identity.physical");
  result.physical_fingerprint = required_string(physical, "fingerprint", 64);
  (void)digest(physical, "fingerprint");
  (void)digest(physical, "model_artifact_fingerprint");
  (void)digest(physical, "tokenizer_fingerprint");
  required_string(physical, "resolved_revision", 256);
  const std::string backend = required_string(physical, "backend", 16);
  const std::string kv = required_string(physical, "kv_format", 64);
  if (!physical.contains("context_allocation") || !physical["context_allocation"].is_number_integer() ||
      !physical.contains("layout_version") || !physical["layout_version"].is_number_integer() ||
      !physical.contains("unified_kv") || !physical["unified_kv"].is_boolean()) {
    throw CacheIdentityError("cache_identity physical descriptor types are invalid");
  }
  if (backend != expected.backend || physical["context_allocation"].get<int64_t>() != expected.context_allocation ||
      kv != expected.kv_format || physical["unified_kv"].get<bool>() != expected.unified_kv ||
      physical["layout_version"].get<int64_t>() != expected.layout_version) {
    throw CacheIdentityError("cache_identity physical model domain does not match loaded runtime");
  }
  return result;
}

}  // namespace clap::llama
